sendrecv/gst-rust/src/main.rs

```diff
@@ -0,0 +1,480 @@
+    ServerRegistered,
+    ServerClosed,
+    PeerConnecting = 3000,
+    PeerConnectionError,
+    PeerConnected,
+    PeerCallNegotiating = 4000,
+    PeerCallStarted,
+    PeerCallError,
+}
+#[derive(Debug)]
+enum Error {
+    GlibError(glib::Error),
+    GlibBoolError(glib::BoolError),
+    GstStateChangeError(gst::StateChangeError),
+}
+
+impl From<glib::Error> for Error {
+    fn from(error: glib::Error) -> Self {
+        Error::GlibError(error)
+    }
+}
+
+impl From<glib::BoolError> for Error {
+    fn from(error: glib::BoolError) -> Self {
+        Error::GlibBoolError(error)
+    }
+}
+
+impl From<gst::StateChangeError> for Error {
+    fn from(error: gst::StateChangeError) -> Self {
+        Error::GstStateChangeError(error)
+    }
+}
+
+const STUN_SERVER: &'static str = "stun-server=stun://stun.l.google.com:19302 ";
+const RTP_CAPS_OPUS: &'static str = "application/x-rtp,media=audio,encoding-name=OPUS,payload=";
+const RTP_CAPS_VP8: &'static str = "application/x-rtp,media=video,encoding-name=VP8,payload=";
+
+fn check_plugins() -> bool {
+    let needed = vec![
+        "opus",
+        "vpx",
+        "nice",
+        "webrtc",
+        "dtls",
+        "srtp",
+        "rtpmanager",
+        "videotestsrc",
+        "audiotestsrc",
+    ];
+    let registry = gst::Registry::get();
+    let mut ret = true;
+    for plugin_name in needed {
+        let plugin = registry.find_plugin(&plugin_name.to_string());
+        if plugin.is_none() {
+            println!("Required gstreamer plugin '{}' not found", plugin_name);
+            ret = false;
+        }
+    }
+    return ret;
+}
+
+fn setup_call(app_control: &Arc<Mutex<AppControl>>) -> AppState {
+    let mut app_control = app_control.lock().unwrap();
+    app_control.app_state = AppState::PeerConnecting;
+    println!(
+        "Setting up signalling server call with {}",
+        app_control.peer_id
+    );
+    app_control
+        .ws_sender
+        .send(format!("SESSION {}", app_control.peer_id))
+        .unwrap();
+    return AppState::PeerConnecting;
+}
+
+fn register_with_server(app_control: &Arc<Mutex<AppControl>>) -> AppState {
+    let mut app_control = app_control.lock().unwrap();
+    app_control.app_state = AppState::ServerRegistering;
+    let our_id = rand::thread_rng().gen_range(10, 10_000);
+    println!("Registering id {} with server", our_id);
+    app_control
+        .ws_sender
+        .send(format!("HELLO {}", our_id))
+        .unwrap();
+    return AppState::ServerRegistering;
+}
+
+fn send_sdp_offer(
+    app_control: &Arc<Mutex<AppControl>>,
+    offer: gstreamer_webrtc::WebRTCSessionDescription,
+) {
+    let app_control = app_control.lock().unwrap();
+    if app_control.app_state < AppState::PeerCallNegotiating {
+        // TODO signal and cleanup
+        panic!("Can't send offer, not in call");
+    };
+    let message = json!({
+      "sdp": {
+        "type": "offer",
+        "sdp": offer.get_sdp().as_text().unwrap(),
+      }
+    });
```

Comment:
For all these JSON things you could define Rust structs, use `serde-derive` and then let `serde` automatically serialize/deserialize them for you instead of having to manually construct/parse JSON. That should make this code a bit more beautiful

Reply:
I got a little held up here because `type` is a reserved word :)
But yes, agreed, will add those

Reply:
```rust
struct Bla {
    #[serde(rename = "type")]
    type_: String,
}
```

Should work

Reply:
I didn't realize initially that an enum can be passed to serde, the fix here is super clean